### DISCUSSION TITLE
multi-tenancy

### DISCUSSION THREAD

**Discussion by jogri** [nhost/nhost#920] [Q&A]
Currently I'm looking into how to achieve a multi-tenancy architecture (database-per-tenant).
It is relatively easy to use only one hasura instance with different databases (same schema) for each tenant.
Now I try to come up with a good strategy for hasura-storage/hasura-auth.
It is probably a good idea to keep the auth service centralised for all tenants - and the permissions for (and isolations of) each database should be solvable. Right now I don't see a good solution for storage: it would be awesome to run only one hasura-storage instance that uses the respective database and storage location per tenant.
Maybe someone can point me to a solution.

**Comment by pkreipke** [comment]
Have you checked out this example? 

https://github.com/nhost/nhost/tree/main/examples/multi-tenant-one-to-many

I've also noticed some topics that might be related such as attaching an "organization" to each user.

**Comment by elitan** [comment]
Maybe if you provide a database model that you plan to use it's easier to reason about.

This is how I would structure a general multi-tenant system (similar to how we do it at [Nhost](http://nhost.io/) with Users and Workspaces.

```
auth.users (Table that already exists in Nhost)
- id
- display_name
- [and more]

storage.files (Table that already exists in Nhost)
- id
- name
- size
- [and more]

workspaces
- id
- name
- created_by_user_id FK to auth.users.id + Hasura preset using `x-hasura-user-id`
- [and more]

workspace_members
- id
- created_at
- workspace_id FK to workspaces.id
- user_id FK to auth.users.id

workspace_files
- id
- file_id FK to storage.files
- workspace_id
- uploaded_by_user_id FK to auth.users.id + Hasura preset using `x-hasura-user-id`
```

## Upload

1. User is signed in and using a specific (current) workspace
2. Upload file
3. Get `file.id` back
4. Insert a new row to `woskpace_files` with `file_id` and the current `workspace_id`.


## Permissions

## Workspaces

To manage the workspace.

Allow to read/update/delete `if workspace.worksapce_members.user_id = x-hasura-user-id` 

## Workspace Members

To manage workspace members.

Allow to insert/read/update/delete `if workspace_members.workspace.worksapce_members.user_id = x-hasura-user-id` 

## Workspace Files

To manage workspace files.

Allow to insert/read/update/delete `if workspace_files.workspace.worksapce_members.user_id = x-hasura-user-id` 

## Files

Allow to insert: always allow

Allow to read/delete `if workspace_files.workspace.worksapce_members.user_id = x-hasura-user-id`

# Summary

Keep everything in the same database and use Hasura's permissions to allow tenants to only manage their own data.

**Comment by jogri** [comment]
Thanks for the replies!
Initially my idea was to have an isolated database and storage location for each tenant (possibly running on their own hardware). Hasura would connect all these databases (same schemas but with different namespaces/prefixes for each tenant) and manage the permissions.

I was hoping to isolate the file storage by tenant as well. The data in storage.files for example would be located in the respective database of each tenant. Same for the files which would be stored respective locations (different tenant servers). So this way there would be one server instance for each tenant but still running only one hasura-storage service.

Example:

**main services server**
- hasura (connecting all tenant databases, managing permission)
- hasura-storage (connecting to the right storage database/buckets)
- hasura-auth
- storage service (connecting to the respective tenant storage location)
- backend services etc.

**server tenant 1**
- database server (tenant 1 data + storage schema)
- location of bucket (tenant 1 files)

**server tenant 2**
- database server (tenant 2 data + storage schema)
- location of bucket (tenant 2 files)

**server tenant 3**
- database server (tenant 3 data + storage schema)
- location of bucket (tenant 3 files)

Then there is also a webserver that runs the rest like the webapp, does all the proxying, monitoring etc. 

Some advantages of this architecture would be:
- best isolation of data
- tenants could run their own servers with database and files (helps with legal responsibilities)
- scalability
- it is possible to move one tenant to a standalone solution (by setting up a "main services server" for a specific tenant)

But I also see potential disadvantages:

- hard to keep all the tenants database schemas in sync (at least it makes it more painful to migrate to new schemas)
- complexity management (e.g. setting up each tenant) and maintenance

But maybe I should give up on that approach (for now) and try the shared database/storage approach.

**Comment by elitan** [comment]
Unless you have **really** strong requirements, I recommend that you do anything in your power to **not** use one database per tenant. 

If you go with one database per tenant, please ensure you fully understand the second-order effects it has in terms of maintenance and complexity for your development team.

> But maybe I should give up on that approach (for now) and try the shared database/storage approach.

I strongly put my vote on this.

**Comment by jogri** [comment]
Thanks elitan. Yeah, I see your point. But it depends on the use case. I expect very view tenants, with large data.
I guess I’ll research a bit more the possibilities on how to isolate data with hasura-storage.  Sorry to ask that here: Is hasura-storage able to use different buckets for the actual data storage (e.g. MinIO)? My current setup uses buckets only on the hasura level but stores the data in the same storage bucket.

**Comment by elitan** [comment]
Yes, it's a complex topic for sure, and there is no general answer that works for everyone.

Hasura Storage uses a single bucket.
